Add dynamic way to choose controller scheme

aswgxf opened the pull request:
Adds an easier way to access data from the parameter server, primarily controller mappings at this point. Closes #8

aswgxf committed:
Merge pull request #1 from MST-Robotics/master

Update

aswgxf committed:
Add way to get information from config files

Adds a way to get information from the included config files and puts them in a struct. Currently buttons must be static casted to an int.

aswgxf committed:
Merge branch 'master' of https://github.com/MST-Robotics/IGVC

aswgxf committed:
Merge branch 'master' of https://github.com/MST-Robotics/IGVC

mfschlittler committed:
Set the doubles for the alternate controller program to ints

  Created a function within the config.cpp that will take the values of the buttons and ensure they are read as ints.

mfschlittler committed:
Merge branch 'master' of https://github.com/aswgxf/IGVC

MSTRobotics committed:
Fix issue with missing line

Stuff now kinda works

aswgxf committed:
Add documentation for config struct and related functions

Added some doxygen documentation for the struct and functions.

aswgxf committed:
Merge branch 'master' of https://github.com/MST-Robotics/IGVC

Fixed conflict and made a couple corrections
Conflicts:
	ros_ws/src/vision/CMakeLists.txt

aswgxf commented:
@matthewia94 @islamelnabarawy

matthewia94 commented on ros_ws/src/common/config.h:
```diff
@@ -0,0 +1,45 @@
+#include "constants.h"
+
+
+struct ConfigValues
```
This struct may end up getting really big when everything is said and done.  A hash table or map might be a better solution.  @islamelnabarawy What do you think?

aswgxf replied:
I'm not sure what those are/how to do that, but I could take a look if it'd be better.

matthewia94 replied:
I'm not sure if it is better or not, lets wait and see what Sam thinks.

matthewia94 commented on ros_ws/src/control/launch/robot.launch:
```diff
@@ -16,6 +16,7 @@
     
     <!-- Load parameter file -->
     <rosparam command="load" file="$(find control)/param/igvc.yaml"/>
+    <rosparam command="load" file="$(find control)/param/controllers/moga.yaml"/>
```
It would be nice if we could pick which controller to use with a command line argument.  We could use the xbox as the default but use a different controller file if they pass it when launching.  More info here: http://wiki.ros.org/roslaunch/XML/arg

matthewia94 commented on ros_ws/src/control/src/control/remote_control.cpp:
```diff
@@ -17,7 +17,7 @@ RemoteControl::RemoteControl()
     cmd_vel.angular.z = 0;
     speed_modifier = 0.5;
-
+    config = initConfigs();
```
I would separate this into its own code block to distinguish it from the rest of the initialization.  It might also be helpful if you left a comment describing what it does here so its not necessary to go looking for its implementation later.

aswgxf replied:
Does it need to be added to  RemoteControl::RemoteControl(std::string twist_topic, std::string joy_topic) ?

matthewia94 replied:
Yes, that would be helpful since that constructor works the same way

matthewia94 commented on ros_ws/src/control/src/control/twist_mux.cpp:
```diff
@@ -11,7 +11,8 @@ TwistMux::TwistMux()
     stopRobot();
-
+    //Get remote values
+    config = initConfigs();
```
Once again I would separate this into blocks.

matthewia94 commented on ros_ws/src/vision/CMakeLists.txt:
```diff
@@ -13,6 +13,7 @@ find_package(catkin REQUIRED COMPONENTS
 ## catkin specific configuration ##
 ###################################
 catkin_package(
```
This should not have been committed as it was fixed in #69.  Please check that you aren't adding unrelated files when adding to your commit.

aswgxf committed:
Set controller through command line argument

Adds a way to set the controller type through a command line argument. The default value is 360.yaml

aswgxf committed:
Seperate initConfigs() from other code

Adds some spacing and a comment for the initConfigs() in twist_mux.cpp and remote_control.cpp

aswgxf committed:
Merge branch 'master' of https://github.com/aswgxf/IGVC

aswgxf committed:
Add initConfigs() to other constructor

matthewia94 commented:
@islamelnabarawy I looked over this and everything seems fine to me.  Would you mind taking a look to make sure that the structure looks okay to you.

islamelnabarawy commented:
@matthewia94 
I agree, the structure would probably end up getting pretty large. There will only ever be a single object of it at any time though, correct? Also, it'll occupy the same size whether we use a dynamic structure or a static one, since all of those configurations have values, right?

If it's true that there will always be exactly one single object of this struct at any time, it might be a good idea to convert it to a singleton pattern with a private constructor. I can elaborate on exactly how that would work if you need me to. I wouldn't do it for this pull request, though; I'd create a separate task for it.

matthewia94 commented on ros_ws/src/control/param/controllers/moga.yaml:
```diff
@@ -0,0 +1,23 @@
+standby_btn:       1
+
+autonomous_btn:    3
+speed_inc_btn:     4
```
These are reversed

aswgxf replied:
Should be fixed now

aswgxf committed:
Fix config file mapping

aswgxf commented:
@matthewia94
